server/src/main/java/org/elasticsearch/search/retriever/CompoundRetrieverBuilder.java

```diff
@@ -50,6 +50,15 @@ public record RetrieverSource(RetrieverBuilder retriever, SearchSourceBuilder so
     protected final List<RetrieverSource> innerRetrievers;
 
     protected CompoundRetrieverBuilder(List<RetrieverSource> innerRetrievers, int rankWindowSize) {
+
+        for (RetrieverSource innerRetriever : innerRetrievers) {
+            if (innerRetriever.retriever.isAllowedAsChildRetriever() == false) {
+                throw new IllegalArgumentException(
+                    "[" + innerRetriever.retriever.getName() + "] must be a standalone or top level retriever only"
```

Comment:
Not sure how clear this terminology will be clear for our users.

Reply:
@kderusso, could you clarify why this restriction was added? Is it related to the window size not being propagated by the parent retriever? Regardless, this is something we should handle natively within the compound retrievers. Reintroducing this kind of restriction would go against the purpose of the recent refactoring, which aimed to eliminate it. All retrievers should be composable at any level.

Reply:
@jimczi I don't think that this is due to the window size but rather how RRF combines the results of the inner retrievers. If `rule` is not at the top level, `rrf` will rank all of the documents that came back from all queries. So if a document that was excluded in the rule retriever was returned in a different retriever it will still be returned, and pinned rank is also not recognized. Furthermore, by the time the query is executed, we don't know anything about the matching rules (so we don't know for example that we should be post-filtering excluded documents). 

This restriction was the best way I could think of to address this issue for RRF, but I'm very open to alternate suggestions! 

Here's an example: 
```
GET my-index/_search?error_trace
{
  "retriever": {
    "rrf": {
      "retrievers": [
        {
          "standard": {
            "query": {
              "query_string": {
                "query": "bar"
              }
            }
          }
        },
        {
          "rule": {
            "match_criteria": {
              "foo": "foo",
              "bar": "bar"
            },
            "ruleset_ids": [
              "my-ruleset"
            ],
            "retriever": {
              "standard": {
                "query": {
                  "match_none": {}
                }
              }
            }
          }
        }
      ]
    }
  },
  "explain": true
}
```


Returns: 

```
{
  "took": 20,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 4,
      "relation": "eq"
    },
    "max_score": 0.016393442,
    "hits": [
      {
        "_shard": "[my-index][0]",
        "_node": "UVqSQ4uqT7CAmfcyrBTxOA",
        "_index": "my-index",
        "_id": "foo",
        "_score": 0.016393442,
        "_source": {
          "text": "foo - pinned doc for foo"
        },
        "_explanation": {
          "value": 0.016393442,
          "description": "rrf score: [0.016393442] computed for initial ranks [0, 1] with rankConstant: [60] as sum of [1 / (rank + rankConstant)] for each query",
          "details": [
            {
              "value": 0,
              "description": "rrf score: [0], result not found in query at index [0]",
              "details": []
            },
            {
              "value": 1,
              "description": "rrf score: [0.016393442], for rank [1] in query at index [1] computed as [1 / (1 + 60)], for matching query with score",
              "details": [
                {
                  "value": 1.7014124e+38,
                  "description": "query rules evaluated rules from rulesets [my-ruleset] and match criteria {bar=bar, foo=foo}",
                  "details": [
                    {
                      "value": 1,
                      "description": "doc [0] with an original score of [1.7014124E38] is at rank [1] from the following source queries.",
                      "details": [
                        {
                          "value": 1.7014124e+38,
                          "description": "sum of:",
                          "details": [
                            {
                              "value": 1.7014124e+38,
                              "description": "max of:",
                              "details": [
                                {
                                  "value": 1.7014124e+38,
                                  "description": "max of:",
                                  "details": [
                                    {
                                      "value": 1.7014124e+38,
                                      "description": "ConstantScore(_id:([7e 8a]))^1.7014124E38",
                                      "details": []
                                    }
                                  ]
                                }
                              ]
                            }
                          ]
                        }
                      ]
                    }
                  ]
                }
              ]
            }
          ]
        }
      },
      {
        "_shard": "[my-index][0]",
        "_node": "UVqSQ4uqT7CAmfcyrBTxOA",
        "_index": "my-index",
        "_id": "bar",
        "_score": 0.016393442,
        "_source": {
          "text": "bar - exclude doc for bar"
        },
        "_explanation": {
          "value": 0.016393442,
          "description": "rrf score: [0.016393442] computed for initial ranks [1, 0] with rankConstant: [60] as sum of [1 / (rank + rankConstant)] for each query",
          "details": [
            {
              "value": 1,
              "description": "rrf score: [0.016393442], for rank [1] in query at index [0] computed as [1 / (1 + 60)], for matching query with score",
              "details": [
                {
                  "value": 1.4021287,
                  "description": "max of:",
                  "details": [
                    {
                      "value": 1.4021287,
                      "description": "weight(text:bar in 0) [PerFieldSimilarity], result of:",
                      "details": [
                        {
                          "value": 1.4021287,
                          "description": "score(freq=2.0), computed as boost * idf * tf from:",
                          "details": [
                            {
                              "value": 2.2,
                              "description": "boost",
                              "details": []
                            },
                            {
                              "value": 1.0296195,
                              "description": "idf, computed as log(1 + (N - n + 0.5) / (n + 0.5)) from:",
                              "details": [
                                {
                                  "value": 2,
                                  "description": "n, number of documents containing term",
                                  "details": []
                                },
                                {
                                  "value": 6,
                                  "description": "N, total number of documents with field",
                                  "details": []
                                }
                              ]
                            },
                            {
                              "value": 0.61899686,
                              "description": "tf, computed as freq / (freq + k1 * (1 - b + b * dl / avgdl)) from:",
                              "details": [
                                {
                                  "value": 2,
                                  "description": "freq, occurrences of term within document",
                                  "details": []
                                },
                                {
                                  "value": 1.2,
                                  "description": "k1, term saturation parameter",
                                  "details": []
                                },
                                {
                                  "value": 0.75,
                                  "description": "b, length normalization parameter",
                                  "details": []
                                },
                                {
                                  "value": 5,
                                  "description": "dl, length of field",
                                  "details": []
                                },
                                {
                                  "value": 4.8333335,
                                  "description": "avgdl, average length of field",
                                  "details": []
                                }
                              ]
                            }
                          ]
                        }
                      ]
                    }
                  ]
                }
              ]
            },
            {
              "value": 0,
              "description": "rrf score: [0], result not found in query at index [1]",
              "details": []
            }
          ]
        }
      },
      {
        "_shard": "[my-index][0]",
        "_node": "UVqSQ4uqT7CAmfcyrBTxOA",
        "_index": "my-index",
        "_id": "bar_no_rule",
        "_score": 0.016129032,
        "_source": {
          "text": "bar_no_rule - bar no rule"
        },
        "_explanation": {
          "value": 0.016129032,
          "description": "rrf score: [0.016129032] computed for initial ranks [2, 0] with rankConstant: [60] as sum of [1 / (rank + rankConstant)] for each query",
          "details": [
            {
              "value": 2,
              "description": "rrf score: [0.016129032], for rank [2] in query at index [0] computed as [1 / (2 + 60)], for matching query with score",
              "details": [
                {
                  "value": 1.1077526,
                  "description": "max of:",
                  "details": [
                    {
                      "value": 1.1077526,
                      "description": "weight(text:bar in 0) [PerFieldSimilarity], result of:",
                      "details": [
                        {
                          "value": 1.1077526,
                          "description": "score(freq=1.0), computed as boost * idf * tf from:",
                          "details": [
                            {
                              "value": 2.2,
                              "description": "boost",
                              "details": []
                            },
                            {
                              "value": 1.0296195,
                              "description": "idf, computed as log(1 + (N - n + 0.5) / (n + 0.5)) from:",
                              "details": [
                                {
                                  "value": 2,
                                  "description": "n, number of documents containing term",
                                  "details": []
                                },
                                {
                                  "value": 6,
                                  "description": "N, total number of documents with field",
                                  "details": []
                                }
                              ]
                            },
                            {
                              "value": 0.48903877,
                              "description": "tf, computed as freq / (freq + k1 * (1 - b + b * dl / avgdl)) from:",
                              "details": [
                                {
                                  "value": 1,
                                  "description": "freq, occurrences of term within document",
                                  "details": []
                                },
                                {
                                  "value": 1.2,
                                  "description": "k1, term saturation parameter",
                                  "details": []
                                },
                                {
                                  "value": 0.75,
                                  "description": "b, length normalization parameter",
                                  "details": []
                                },
                                {
                                  "value": 4,
                                  "description": "dl, length of field",
                                  "details": []
                                },
                                {
                                  "value": 4.8333335,
                                  "description": "avgdl, average length of field",
                                  "details": []
                                }
                              ]
                            }
                          ]
                        }
                      ]
                    }
                  ]
                }
              ]
            },
            {
              "value": 0,
              "description": "rrf score: [0], result not found in query at index [1]",
              "details": []
            }
          ]
        }
      },
      {
        "_shard": "[my-index][0]",
        "_node": "UVqSQ4uqT7CAmfcyrBTxOA",
        "_index": "my-index",
        "_id": "foo2",
        "_score": 0.016129032,
        "_source": {
          "text": "foo2 - 2nd pinned doc for foo"
        },
        "_explanation": {
          "value": 0.016129032,
          "description": "rrf score: [0.016129032] computed for initial ranks [0, 2] with rankConstant: [60] as sum of [1 / (rank + rankConstant)] for each query",
          "details": [
            {
              "value": 0,
              "description": "rrf score: [0], result not found in query at index [0]",
              "details": []
            },
            {
              "value": 2,
              "description": "rrf score: [0.016129032], for rank [2] in query at index [1] computed as [1 / (2 + 60)], for matching query with score",
              "details": [
                {
                  "value": 1.7014122e+38,
                  "description": "query rules evaluated rules from rulesets [my-ruleset] and match criteria {bar=bar, foo=foo}",
                  "details": [
                    {
                      "value": 2,
                      "description": "doc [5] with an original score of [1.7014122E38] is at rank [2] from the following source queries.",
                      "details": [
                        {
                          "value": 1.7014122e+38,
                          "description": "sum of:",
                          "details": [
                            {
                              "value": 1.7014122e+38,
                              "description": "max of:",
                              "details": [
                                {
                                  "value": 1.7014122e+38,
                                  "description": "max of:",
                                  "details": [
                                    {
                                      "value": 1.7014122e+38,
                                      "description": "ConstantScore(_id:([7e 8a 36]))^1.7014122E38",
                                      "details": []
                                    }
                                  ]
                                }
                              ]
                            }
                          ]
                        }
                      ]
                    }
                  ]
                }
              ]
            }
          ]
        }
      }
    ]
  }
}
```

Reply:
> This restriction was the best way I could think of to address this issue for RRF, but I'm very open to alternate suggestions!

But what you describe is the correct behaviour. If the rules retriever is part of the sub-retrievers in RRF then the rules should only apply to the sub-retriever level. If the intent is to apply the rules on the final top hits then the retriever should be placed at the root. Using a rules retriever as a sub-retriever shouldn't be the main use case but I also don't see why it should be forbidden. It works as expected in the current framework and this could become more useful as we add compound retrievers in the framework.

Reply:
I really dislike this from a usability perspective. I think that this is going to be extremely opaque and hard to understand for users, and generate a lot of questions for support/community. However, I can remove it for now.